tools/fuzzer/expr_gen.cc

```diff
@@ -0,0 +1,149 @@
+  std::uniform_int_distribution<uint64_t> value_distr(0, MAX_INT_VALUE);
+  auto value = value_distr(m_rng);
+  auto gen_parens = fifty_fifty();
+
+  return IntegerConstant(value, gen_parens);
+}
+
+DoubleConstant ExprGenerator::gen_double_constant(const WeightsArray&) {
+  std::uniform_real_distribution<double> value_distr(0.0, MAX_DOUBLE_VALUE);
+  auto value = value_distr(m_rng);
+  auto gen_parens = fifty_fifty();
+
+  return DoubleConstant(value, gen_parens);
+}
+
+VariableExpr ExprGenerator::gen_variable_expr(const WeightsArray&) {
+  auto gen_parens = fifty_fifty();
+
+  return VariableExpr(VAR, gen_parens);
+}
+
+BinaryExpr ExprGenerator::gen_binary_expr(const WeightsArray& weights) {
+  WeightsArray new_weights = weights;
+
+  auto idx = (size_t) ExprKind::BinaryExpr;
+  new_weights[idx] *= EXPR_KIND_INFO[idx].dampening_factor;
+
+  auto lhs = gen_with_weights(new_weights);
+  auto rhs = gen_with_weights(new_weights);
+
+  // The interval of `uniform_int_distribution` is inclusive.
+  std::uniform_int_distribution<size_t> op_distr(0, (size_t) BinOp::EnumLast - 1);
+  auto op = (BinOp) op_distr(m_rng);
+  bool gen_parens = fifty_fifty();
+
+  return BinaryExpr(std::move(lhs), op, std::move(rhs), gen_parens);
+}
+
+UnaryExpr ExprGenerator::gen_unary_expr(const WeightsArray& weights) {
+  WeightsArray new_weights = weights;
+
+  auto idx = (size_t) ExprKind::UnaryExpr;
+  new_weights[idx] *= EXPR_KIND_INFO[idx].dampening_factor;
+
+  auto expr = gen_with_weights(new_weights);
+
+  std::uniform_int_distribution<size_t> op_distr(0, (size_t) UnOp::EnumLast - 1);
+  auto op = (UnOp) op_distr(m_rng);
+  bool gen_parens = fifty_fifty();
+
+  return UnaryExpr(op, std::move(expr), gen_parens);
+}
+
+Expr ExprGenerator::gen_with_weights(const WeightsArray& weights) {
+  // No need to pull `<numeric>` just for one sum.
+  float sum = 0;
+  for (const auto& e: weights) {
+    sum += e;
+  }
+
+  std::uniform_real_distribution<float> distr(0, sum);
+  auto val = distr(m_rng);
+
+  // Make sure `kind` is always initialised
+  float running_sum = 0;
+  ExprKind kind = ExprKind::IntegerConstant;;
+  for (size_t i = 0; i < weights.size(); i++) {
+    running_sum += weights[i];
+    if (val < running_sum) {
+      kind = (ExprKind) i;
+      break;
+    }
+  }
+
+  switch (kind) {
+    case ExprKind::IntegerConstant:
+      return gen_integer_constant(weights);
+
+    case ExprKind::DoubleConstant:
+      return gen_double_constant(weights);
+
+    case ExprKind::VariableExpr:
+      return gen_variable_expr(weights);
+
+    case ExprKind::BinaryExpr:
+      return gen_binary_expr(weights);
+
+    case ExprKind::UnaryExpr:
+      return gen_unary_expr(weights);
+
+    default:
+      assert(false && "Unreachable");
+  }
+}
+
+Expr ExprGenerator::generate() {
+  WeightsArray weights;
+  for (size_t i = 0; i < weights.size(); i++) {
+    weights[i] = EXPR_KIND_INFO[i].initial_weight;
+  }
```

Comment:
`ExprGenerator` looks basically like a "reverse" parser. It implements the grammar directly, by manually handling each production branch. This a valid approach, but I'm not so sure how it will scale. We already have one grammar implementation (lldb-eval parser), maintaining two might be a hassle.

My initial idea was something like this: `ExprGenerator` accepts ebnf grammar as a parameter, builds a production table and then generates the expression from it. This way we don't need to manually implement all the production rules and (very important) don't have to change the fuzzer when we change the grammar and parser. We can still have some customizations and extensions (e.g. we know that `type_id` is a type and instead of generating it from grammar we can pick a random one from the binary), but even "bare-bone" generator should provide valid (grammar-wise) expressions.